tests/system/test_dbapi.py

```diff
@@ -425,6 +425,125 @@ def test_read_timestamp_client_side_autocommit(self):
+                    (2, "first-name-2", "last-name-2", "test.email@domen.ru"),
+                    (3, "first-name-3", "last-name-3", "test.email@domen.ru"),
+                    (4, "first-name-4", "last-name-4", "test.email@domen.ru"),
+                    (5, "first-name-5", "last-name-5", "test.email@domen.ru"),
+                    (6, "first-name-6", "last-name-6", "test.email@domen.ru"),
+                ]
+            ).sort()
+        )
+
+        # Test starting another dml batch in same connection post commit works
+        self._cursor.execute("start batch dml")
+        self._insert_row(7)
+        self._insert_row(8)
+        self._cursor.execute("run batch")
+
+        self._insert_row(9)
+
+        if not auto_commit:
+            self._conn.commit()
+
+        self._cursor.execute("SELECT * FROM contacts")
+        assert len(self._cursor.fetchall()) == 9
+
+    def test_abort_batch_dml(self):
+        """Test abort batch dml."""
+
+        self._cursor.execute("start batch dml")
+        self._insert_row(1)
+        self._insert_row(2)
+        self._cursor.execute("abort batch")
+
+        self._insert_row(3)
+        self._conn.commit()
+
+        self._cursor.execute("SELECT * FROM contacts")
+        got_rows = self._cursor.fetchall()
+        assert len(got_rows) == 1
+        assert got_rows == [(3, "first-name-3", "last-name-3", "test.email@domen.ru")]
+
+    def test_batch_dml_invalid_statements(self):
+        """Test batch dml having invalid statements."""
+
+        # Test first statement in batch is invalid
+        self._cursor.execute("start batch dml")
+        self._cursor.execute(
+            """
+            INSERT INTO unknown_table (contact_id, first_name, last_name, email)
+            VALUES (2, 'first-name', 'last-name', 'test.email@domen.ru')
+            """
+        )
+        self._insert_row(1)
+        self._insert_row(2)
+        with pytest.raises(OperationalError):
+            self._cursor.execute("run batch")
+
+        # Test middle statement in batch is invalid
+        self._cursor.execute("start batch dml")
+        self._insert_row(1)
+        self._cursor.execute(
+            """
+            INSERT INTO unknown_table (contact_id, first_name, last_name, email)
+            VALUES (2, 'first-name', 'last-name', 'test.email@domen.ru')
+            """
+        )
+        self._insert_row(2)
+        with pytest.raises(OperationalError):
+            self._cursor.execute("run batch")
+
+        # Test last statement in batch is invalid
+        self._cursor.execute("start batch dml")
+        self._insert_row(1)
+        self._insert_row(2)
+        self._cursor.execute(
+            """
+            INSERT INTO unknown_table (contact_id, first_name, last_name, email)
+            VALUES (2, 'first-name', 'last-name', 'test.email@domen.ru')
+            """
+        )
+        with pytest.raises(OperationalError):
+            self._cursor.execute("run batch")
```

Comment:
I would have expected this to return a specific batch error that:
1. Contains the error code, message etc.
2. And contains the update counts of the statements that did succeed.

See for example this for the JDBC driver: https://github.com/googleapis/java-spanner-jdbc/blob/1f89f78c37b9e118e2c0cbc7f56d3eb1d5745863/src/test/java/com/google/cloud/spanner/jdbc/it/ITJdbcPreparedStatementTest.java#L797

Reply:
As discussed we need to create a custom exception class, so will take it in follow up PR